File: Data Structures & Algorithms/reverse-a-linked-list/submission-2.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
    ListNode* reverseList(ListNode* head) {
        if (!head) { return head; }
        
        ListNode* prev = nullptr;
        ListNode* curr = head;
        ListNode* temp;

        while (curr) {
            temp = curr->next;
            curr->next = prev;
            prev = curr;
            curr = temp;
        }

        return prev;
    }
};

/*
    0. make a pointer to the head
    1. push every element of the list onto a stack
    2. loop from the back and assign every element to the list
    3. return the new list
*/

File: Data Structures & Algorithms/merge-two-sorted-linked-lists/submission-2.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def mergeTwoLists(self, list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:
        """
            0. if both empty, return None
            0b. if one is empty, return the other
            1. init head with the smallest, update ptrs accordingly
            2. while list1 and list2
            3. if list1.val < list2.val:
                init temp node to list1
                head.next = temp
                head = head.next
                list1 = list1.next
            4. else, repeat prev for list2
            5. while list1, add list1 nodes to head
            6. while list2, add list2 nodes to head ptr
            7. return head
        """
        if not list1 and not list2:
            return list1
        if not list1:
            return list2
        if not list2:
            return list1
        
        head = ListNode()
        if list1.val < list2.val:
            head.val = list1.val
            list1 = list1.next
        else:
            head.val = list2.val
            list2 = list2.next
        curr = head
        while list1 and list2:
            temp = ListNode()
            if list1.val < list2.val:
                temp.val = list1.val
                list1 = list1.next
            else:
                temp.val = list2.val
                list2 = list2.next
            curr.next = temp
            curr = curr.next
        while list1:
            temp = ListNode()
            temp.val = list1.val
            list1 = list1.next
            curr.next = temp
            curr = curr.next
        while list2:
            temp = ListNode()
            temp.val = list2.val
            list2 = list2.next
            curr.next = temp
            curr = curr.next
        return head
        